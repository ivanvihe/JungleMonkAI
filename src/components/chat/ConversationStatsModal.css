.conversation-stats {
  display: flex;
  flex-direction: column;
  gap: 18px;
}

.conversation-stats__summary {
  display: flex;
  gap: 16px;
  flex-wrap: wrap;
}

.conversation-stats__total,
.conversation-stats__pending {
  flex: 1 1 180px;
  display: flex;
  flex-direction: column;
  gap: 6px;
  padding: 16px;
  border-radius: 16px;
  background: rgba(255, 255, 255, 0.04);
  border: 1px solid rgba(255, 255, 255, 0.08);
}

.conversation-stats__total {
  background: linear-gradient(135deg, rgba(142, 141, 255, 0.28), rgba(77, 208, 225, 0.18));
  border-color: rgba(142, 141, 255, 0.4);
}

.conversation-stats__label {
  font-size: 11px;
  text-transform: uppercase;
  letter-spacing: 0.6px;
  color: rgba(255, 255, 255, 0.65);
}

.conversation-stats__value {
  font-size: 28px;
  font-weight: 600;
  letter-spacing: 0.5px;
}

.conversation-stats__value.is-warning {
  color: rgba(255, 183, 77, 0.95);
}

.conversation-stats__hint {
  font-size: 13px;
  color: rgba(255, 255, 255, 0.7);
}

.conversation-stats__grid {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(160px, 1fr));
  gap: 12px;
}

.conversation-stats__card {
  display: flex;
  flex-direction: column;
  gap: 4px;
  padding: 14px 16px;
  border-radius: 14px;
  background: rgba(255, 255, 255, 0.05);
  border: 1px solid rgba(255, 255, 255, 0.08);
}

.conversation-stats__card-label {
  font-size: 12px;
  letter-spacing: 0.4px;
  color: rgba(255, 255, 255, 0.65);
}

.conversation-stats__card-value {
  font-size: 20px;
  font-weight: 600;
}

.conversation-stats__meta {
  display: flex;
  flex-direction: column;
  gap: 12px;
  padding: 14px 16px;
  border-radius: 14px;
  background: rgba(255, 255, 255, 0.04);
  border: 1px solid rgba(255, 255, 255, 0.08);
}

.conversation-stats__meta-row {
  display: flex;
  justify-content: space-between;
  gap: 12px;
  flex-wrap: wrap;
}

.conversation-stats__meta-label {
  font-size: 12px;
  text-transform: uppercase;
  letter-spacing: 0.5px;
  color: rgba(255, 255, 255, 0.6);
}

.conversation-stats__meta-value {
  display: inline-flex;
  align-items: center;
  gap: 8px;
  font-size: 13px;
  color: rgba(255, 255, 255, 0.85);
}

.conversation-stats__meta-author {
  display: inline-flex;
  align-items: center;
  gap: 6px;
  padding: 4px 10px;
  border-radius: 999px;
  background: rgba(255, 255, 255, 0.08);
  font-size: 11px;
  text-transform: uppercase;
  letter-spacing: 0.6px;
}

.conversation-stats__jarvis {
  display: flex;
  flex-direction: column;
  gap: 14px;
  padding: 16px;
  border-radius: 14px;
  background: rgba(255, 255, 255, 0.04);
  border: 1px solid rgba(255, 255, 255, 0.08);
}

.conversation-stats__section-header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  gap: 12px;
}

.conversation-stats__section-title {
  font-size: 13px;
  letter-spacing: 0.6px;
  text-transform: uppercase;
  color: rgba(255, 255, 255, 0.75);
}

.status-pill {
  display: inline-flex;
  align-items: center;
  gap: 6px;
  padding: 4px 10px;
  border-radius: 999px;
  font-size: 11px;
  font-weight: 600;
  letter-spacing: 0.5px;
  text-transform: uppercase;
  background: rgba(255, 255, 255, 0.08);
  color: rgba(255, 255, 255, 0.85);
}

.status-pill--success {
  background: rgba(52, 211, 153, 0.18);
  color: #d1fae5;
}

.status-pill--warning {
  background: rgba(251, 191, 36, 0.22);
  color: #fde68a;
}

.status-pill--danger {
  background: rgba(248, 113, 113, 0.25);
  color: #fecaca;
}

.status-pill--muted {
  background: rgba(148, 163, 184, 0.18);
  color: #e2e8f0;
}

.conversation-stats__jarvis-grid {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(140px, 1fr));
  gap: 12px;
}

.conversation-stats__jarvis-item {
  display: flex;
  flex-direction: column;
  gap: 4px;
  padding: 10px 12px;
  border-radius: 12px;
  background: rgba(255, 255, 255, 0.03);
  border: 1px solid rgba(255, 255, 255, 0.08);
}

.conversation-stats__downloads {
  display: flex;
  flex-direction: column;
  gap: 6px;
}

.conversation-stats__download-list {
  list-style: none;
  margin: 0;
  padding: 0;
  display: flex;
  flex-direction: column;
  gap: 6px;
}

.conversation-stats__download-item {
  display: flex;
  justify-content: space-between;
  align-items: center;
  gap: 12px;
  padding: 8px 10px;
  border-radius: 10px;
  background: rgba(255, 255, 255, 0.05);
}

.conversation-stats__download-name {
  font-size: 13px;
  color: rgba(255, 255, 255, 0.88);
}

.conversation-stats__download-progress {
  font-size: 12px;
  font-weight: 600;
  color: rgba(255, 255, 255, 0.78);
}

.conversation-stats__downloads-empty {
  font-size: 12px;
  color: rgba(255, 255, 255, 0.6);
}

.conversation-stats__jarvis-action {
  align-self: flex-start;
  padding: 8px 16px;
  border-radius: 999px;
  border: 1px solid rgba(255, 255, 255, 0.2);
  background: rgba(255, 255, 255, 0.08);
  color: rgba(255, 255, 255, 0.92);
  font-size: 12px;
  text-transform: uppercase;
  letter-spacing: 0.8px;
  cursor: pointer;
  transition: background 0.2s ease, border 0.2s ease, transform 0.2s ease;
}

.conversation-stats__jarvis-action:hover:not(:disabled) {
  transform: translateY(-1px);
  background: rgba(142, 141, 255, 0.25);
  border-color: rgba(142, 141, 255, 0.5);
}

.conversation-stats__jarvis-action:disabled {
  opacity: 0.6;
  cursor: progress;
}
