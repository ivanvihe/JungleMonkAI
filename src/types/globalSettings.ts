import type { AgentManifestCache } from './agents';

export type BuiltinProvider = 'openai' | 'anthropic' | 'groq' | 'github' | 'gitlab';

export type SupportedProvider = BuiltinProvider | (string & {});

export const BUILTIN_PROVIDERS: BuiltinProvider[] = ['openai', 'anthropic', 'groq', 'github', 'gitlab'];

export type ApiKeySettings = Record<string, string>;

export interface CommandPresetSettings {
  temperature?: number;
  maxTokens?: number;
}

export interface CommandPreset {
  id: string;
  label: string;
  prompt: string;
  description?: string;
  provider?: string;
  model?: string;
  settings?: CommandPresetSettings;
}

export interface RoutingRule {
  provider: string;
  model: string;
  commandPresetId?: string;
}

export type DefaultRoutingRules = Record<string, RoutingRule>;

export interface GlobalSettings {
  version: number;
  apiKeys: ApiKeySettings;
  commandPresets: CommandPreset[];
  defaultRoutingRules: DefaultRoutingRules;
  enabledPlugins: string[];
  approvedManifests: AgentManifestCache;
}
