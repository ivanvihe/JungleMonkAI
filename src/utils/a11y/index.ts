export * from './focus';
export * from './aria';
