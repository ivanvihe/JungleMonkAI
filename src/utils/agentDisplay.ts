import { AgentDefinition } from '../core/agents/agentRegistry';

export const getAgentDisplayName = (agent: AgentDefinition): string =>
  agent.kind === 'local' ? 'Jarvis' : agent.name;

export const getAgentVersionLabel = (agent: AgentDefinition): string =>
  agent.kind === 'local' ? agent.name : agent.name;

export const getAgentChannelLabel = (agent: AgentDefinition): string => {
  if (agent.kind === 'local') {
    return 'Jarvis';
  }

  if (agent.channel === 'gpt') {
    return 'GPT';
  }

  if (agent.channel === 'claude') {
    return 'Claude';
  }

  if (agent.channel === 'groq') {
    return 'Groq';
  }

  return agent.name;
};
