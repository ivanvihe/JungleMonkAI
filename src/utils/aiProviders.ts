export interface ChatProviderRequest {
  apiKey: string;
  model: string;
  prompt: string;
  systemPrompt?: string;
  maxTokens?: number;
  temperature?: number;
}

const DEFAULT_MAX_TOKENS = 1024;
const DEFAULT_TEMPERATURE = 0.7;

const resolveMessage = (value: unknown): string => {
  if (!value) {
    return '';
  }
  if (typeof value === 'string') {
    return value;
  }
  if (Array.isArray(value)) {
    return value
      .map(entry => {
        if (!entry) {
          return '';
        }
        if (typeof entry === 'string') {
          return entry;
        }
        if (typeof entry === 'object' && 'text' in entry) {
          const text = (entry as { text?: string }).text;
          return typeof text === 'string' ? text : '';
        }
        return '';
      })
      .filter(Boolean)
      .join('\n');
  }
  if (typeof value === 'object' && 'content' in (value as Record<string, unknown>)) {
    return resolveMessage((value as Record<string, unknown>).content);
  }
  return '';
};

const ensureContent = (raw: string, provider: string): string => {
  const content = raw.trim();
  if (!content) {
    throw new Error(`${provider} devolvió una respuesta vacía`);
  }
  return content;
};

export const callOpenAIChat = async ({
  apiKey,
  model,
  prompt,
  systemPrompt,
  maxTokens = DEFAULT_MAX_TOKENS,
  temperature = DEFAULT_TEMPERATURE,
}: ChatProviderRequest): Promise<string> => {
  const response = await fetch('https://api.openai.com/v1/chat/completions', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      model,
      max_tokens: maxTokens,
      temperature,
      messages: [
        ...(systemPrompt
          ? [
              {
                role: 'system',
                content: systemPrompt,
              },
            ]
          : []),
        {
          role: 'user',
          content: prompt,
        },
      ],
    }),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error?.error?.message || `OpenAI respondió con ${response.status}`);
  }

  const payload = await response.json();
  const choice = payload?.choices?.[0]?.message?.content;
  return ensureContent(resolveMessage(choice), 'OpenAI');
};

export const callGroqChat = async ({
  apiKey,
  model,
  prompt,
  systemPrompt,
  maxTokens = DEFAULT_MAX_TOKENS,
  temperature = DEFAULT_TEMPERATURE,
}: ChatProviderRequest): Promise<string> => {
  const response = await fetch('https://api.groq.com/openai/v1/chat/completions', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      model,
      max_tokens: maxTokens,
      temperature,
      messages: [
        ...(systemPrompt
          ? [
              {
                role: 'system',
                content: systemPrompt,
              },
            ]
          : []),
        {
          role: 'user',
          content: prompt,
        },
      ],
    }),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error?.error?.message || `Groq respondió con ${response.status}`);
  }

  const payload = await response.json();
  const choice = payload?.choices?.[0]?.message?.content;
  return ensureContent(resolveMessage(choice), 'Groq');
};

export const callAnthropicChat = async ({
  apiKey,
  model,
  prompt,
  systemPrompt,
  maxTokens = DEFAULT_MAX_TOKENS,
  temperature = DEFAULT_TEMPERATURE,
}: ChatProviderRequest): Promise<string> => {
  const response = await fetch('https://api.anthropic.com/v1/messages', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-api-key': apiKey,
      'anthropic-version': '2023-06-01',
    },
    body: JSON.stringify({
      model,
      max_tokens: maxTokens,
      temperature,
      ...(systemPrompt ? { system: systemPrompt } : {}),
      messages: [
        {
          role: 'user',
          content: prompt,
        },
      ],
    }),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    const message = error?.error?.message || error?.error || error?.message;
    throw new Error(message || `Anthropic respondió con ${response.status}`);
  }

  const payload = await response.json();
  const content = payload?.content;
  return ensureContent(resolveMessage(content), 'Anthropic');
};
