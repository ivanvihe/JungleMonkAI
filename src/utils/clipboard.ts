export const copyToClipboard = async (value: string): Promise<void> => {
  if (typeof navigator !== 'undefined' && navigator?.clipboard?.writeText) {
    await navigator.clipboard.writeText(value);
    return;
  }

  // Fallback for environments sin API de clipboard (por ejemplo tests)
  if (typeof document === 'undefined') {
    return;
  }

  const textarea = document.createElement('textarea');
  textarea.value = value;
  textarea.setAttribute('readonly', 'true');
  textarea.style.position = 'absolute';
  textarea.style.left = '-9999px';
  document.body.appendChild(textarea);
  textarea.focus();
  textarea.select();
  document.execCommand('copy');
  document.body.removeChild(textarea);
};

export default copyToClipboard;
