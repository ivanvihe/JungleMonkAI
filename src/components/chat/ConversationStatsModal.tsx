import React, { useMemo } from 'react';
import './ConversationStatsModal.css';
import { OverlayModal } from '../common/OverlayModal';
import { useMessages } from '../../core/messages/MessageContext';
import { ChatAuthor } from '../../core/messages/messageTypes';

interface ConversationStatsModalProps {
  isOpen: boolean;
  onClose: () => void;
}

interface AuthorBreakdown {
  author: ChatAuthor;
  count: number;
  label: string;
}

export const ConversationStatsModal: React.FC<ConversationStatsModalProps> = ({ isOpen, onClose }) => {
  const { messages, pendingResponses, formatTimestamp } = useMessages();

  const stats = useMemo(() => {
    const publicMessages = messages.filter(message => message.visibility !== 'internal');

    const breakdown: AuthorBreakdown[] = (
      [
        { author: 'user' as const, label: 'Mensajes de usuario' },
        { author: 'agent' as const, label: 'Respuestas de agentes' },
        { author: 'system' as const, label: 'Mensajes del sistema' },
      ] satisfies { author: ChatAuthor; label: string }[]
    ).map(({ author, label }) => ({
      author,
      label,
      count: publicMessages.filter(message => message.author === author).length,
    }));

    const lastMessage = messages[messages.length - 1] ?? null;

    return {
      totalMessages: messages.length,
      publicCount: publicMessages.length,
      internalCount: messages.length - publicMessages.length,
      breakdown,
      pendingResponses,
      lastMessage,
    };
  }, [messages, pendingResponses]);

  const lastTimestamp = stats.lastMessage ? formatTimestamp(stats.lastMessage.timestamp) : null;
  const lastDate = stats.lastMessage ? new Date(stats.lastMessage.timestamp) : null;
  const lastAbsoluteTimestamp =
    lastDate && !Number.isNaN(lastDate.getTime()) ? lastDate.toLocaleString() : null;

  return (
    <OverlayModal
      title="Estadísticas de la conversación"
      isOpen={isOpen}
      onClose={onClose}
      width={420}
    >
      <div className="conversation-stats" aria-live="polite">
        <div className="conversation-stats__summary">
          <div className="conversation-stats__total">
            <span className="conversation-stats__label">Mensajes totales</span>
            <span className="conversation-stats__value">{stats.totalMessages}</span>
            <span className="conversation-stats__hint">
              {stats.publicCount} públicos · {stats.internalCount} internos
            </span>
          </div>
          <div className="conversation-stats__pending">
            <span className="conversation-stats__label">Pendientes</span>
            <span
              className={`conversation-stats__value${stats.pendingResponses ? ' is-warning' : ''}`}
            >
              {stats.pendingResponses}
            </span>
          </div>
        </div>

        <div className="conversation-stats__grid">
          {stats.breakdown.map(bucket => (
            <div key={bucket.author} className="conversation-stats__card">
              <span className="conversation-stats__card-label">{bucket.label}</span>
              <span className="conversation-stats__card-value">{bucket.count}</span>
            </div>
          ))}
        </div>

        <div className="conversation-stats__meta">
          <div className="conversation-stats__meta-row">
            <span className="conversation-stats__meta-label">Último mensaje</span>
            <span className="conversation-stats__meta-value" title={lastAbsoluteTimestamp ?? undefined}>
              {stats.lastMessage ? (
                <>
                  <span className="conversation-stats__meta-author">{stats.lastMessage.author}</span>
                  <span>{lastTimestamp}</span>
                </>
              ) : (
                'Sin actividad registrada'
              )}
            </span>
          </div>
        </div>
      </div>
    </OverlayModal>
  );
};

export default ConversationStatsModal;
