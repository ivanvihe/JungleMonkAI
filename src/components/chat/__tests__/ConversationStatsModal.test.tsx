import React from 'react';
import { describe, expect, it, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/vitest';
import { ConversationStatsModal } from '../ConversationStatsModal';
import type { JarvisCoreContextValue } from '../../../core/jarvis/JarvisCoreContext';
import { DEFAULT_GLOBAL_SETTINGS } from '../../../utils/globalSettings';

const useMessagesMock = vi.fn();
const useJarvisCoreMock = vi.fn<JarvisCoreContextValue, []>();

vi.mock('../../../core/messages/MessageContext', () => ({
  useMessages: () => useMessagesMock(),
}));

vi.mock('../../../core/jarvis/JarvisCoreContext', async () => {
  const actual = await vi.importActual<typeof import('../../../core/jarvis/JarvisCoreContext')>(
    '../../../core/jarvis/JarvisCoreContext',
  );
  return {
    ...actual,
    useJarvisCore: () => useJarvisCoreMock(),
  };
});

describe('ConversationStatsModal', () => {
  it('muestra el estado de Jarvis Core y permite reintentar la conexión', () => {
    const now = new Date().toISOString();
    const later = new Date(Date.now() + 1000).toISOString();
    useMessagesMock.mockReturnValue({
      messages: [
        { id: '1', author: 'user', content: 'hola', timestamp: now, visibility: 'public' },
        { id: '2', author: 'agent', content: 'respuesta', timestamp: later, visibility: 'public' },
      ],
      pendingResponses: 1,
      formatTimestamp: (value: number) => new Date(value).toLocaleTimeString('es-ES'),
    });

    const ensureOnline = vi.fn().mockResolvedValue(undefined);
    useJarvisCoreMock.mockReturnValue({
      connected: true,
      lastError: null,
      activeModel: 'phi-2',
      downloads: {
        phi: { modelId: 'phi', status: 'downloading', downloaded: 512, total: 1024, percent: 50, error: null, errorCode: null },
      },
      models: [],
      runtimeStatus: 'ready',
      uptimeMs: 12_000,
      config: DEFAULT_GLOBAL_SETTINGS.jarvisCore,
      baseUrl: 'http://127.0.0.1:8000',
      lastHealthMessage: 'ok',
      ensureOnline,
      refreshModels: vi.fn(),
      downloadModel: vi.fn(),
      activateModel: vi.fn(),
      invokeChat: vi.fn(),
      launchAction: vi.fn(),
    });

    const { container } = render(<ConversationStatsModal isOpen onClose={vi.fn()} />);

    expect(screen.getByText('JarvisCore')).toBeInTheDocument();
    expect(screen.getByText(/phi-2/)).toBeInTheDocument();
    expect(screen.getByText(/50%/)).toBeInTheDocument();

    const button = screen.getByRole('button', { name: /Reintentar conexión/i });
    button.click();
    expect(ensureOnline).toHaveBeenCalled();

    expect(container.querySelector('.conversation-stats__jarvis')).toMatchSnapshot('jarvis-section');
  });
});
