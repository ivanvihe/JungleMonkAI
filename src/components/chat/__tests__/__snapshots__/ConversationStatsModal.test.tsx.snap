// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConversationStatsModal > muestra el estado de Jarvis Core y permite reintentar la conexión > jarvis-section 1`] = `
<div
  aria-label="Estado de Jarvis Core"
  class="conversation-stats__jarvis"
  role="region"
>
  <div
    class="conversation-stats__section-header"
  >
    <span
      class="conversation-stats__section-title"
    >
      JarvisCore
    </span>
    <span
      class="status-pill status-pill--success"
      title="ok"
    >
      Operativo
    </span>
  </div>
  <div
    class="conversation-stats__jarvis-grid"
  >
    <div
      class="conversation-stats__jarvis-item"
    >
      <span
        class="conversation-stats__meta-label"
      >
        Host
      </span>
      <span
        class="conversation-stats__meta-value"
      >
        127.0.0.1
      </span>
    </div>
    <div
      class="conversation-stats__jarvis-item"
    >
      <span
        class="conversation-stats__meta-label"
      >
        Puerto
      </span>
      <span
        class="conversation-stats__meta-value"
      >
        8000
      </span>
    </div>
    <div
      class="conversation-stats__jarvis-item"
    >
      <span
        class="conversation-stats__meta-label"
      >
        Modelo activo
      </span>
      <span
        class="conversation-stats__meta-value"
      >
        phi-2
      </span>
    </div>
    <div
      class="conversation-stats__jarvis-item"
    >
      <span
        class="conversation-stats__meta-label"
      >
        Actividad
      </span>
      <span
        class="conversation-stats__meta-value"
      >
        0m 12s
      </span>
    </div>
  </div>
  <div
    class="conversation-stats__downloads"
  >
    <span
      class="conversation-stats__meta-label"
    >
      Descargas en curso
    </span>
    <ul
      class="conversation-stats__download-list"
    >
      <li
        class="conversation-stats__download-item"
      >
        <span
          class="conversation-stats__download-name"
        >
          phi
        </span>
        <span
          class="conversation-stats__download-progress"
        >
          50%
        </span>
      </li>
    </ul>
  </div>
  <button
    class="conversation-stats__jarvis-action"
    type="button"
  >
    Reintentar conexión
  </button>
</div>
`;
